[#22] 주문관련 비즈니스 로직

jjeda opened the pull request:
- 주문 관련 domain Dto 생성
- 주문 레포지토리 생성
- 주문 서비스 생성
- 주문 생성 & 취소 메서드
- 주문 컨트롤러
- 주문 통합 테스트

jjeda committed:
[#22] 주문관련 비즈니스 로직

- 주문 관련 domain Dto 생성
- 주문 레포지토리 생성
- 주문 서비스 생성
- 주문 생성 & 취소 메서드

jjeda committed:
[#22] 주문 비즈니스 로직 수정

- 주문 서비스 수정
- 주문 컨트롤러 추가

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/controller/AdminOrderController.java:
```diff
@@ -0,0 +1,32 @@
+
+    private final OrderService orderService;
+
+    @GetMapping("/{orderId}/delivery/start")
```
상태의 일부를 변경하는거니 `PatchMapping`이 맞는 것 같네요

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/controller/AdminOrderController.java:
```diff
@@ -0,0 +1,32 @@
+    private final OrderService orderService;
+
+    @GetMapping("/{orderId}/delivery/start")
+    public ResponseEntity startDelivery(@PathVariable Long orderId) {
```
이런 행동이 정적인 API는 유지보수성에 좋지 않습니다.
status가 늘어날수록 엔드포인트를 계속 추가해줘야하고, 클라이언트는 해당 URL을 계속 추가해줘야합니다.
아래의 finish도 마찬가지입니다~

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/controller/AdminOrderController.java:
```diff
@@ -0,0 +1,32 @@
+@Controller
+@RequiredArgsConstructor
+@RequestMapping("/admin/orders")
+public class AdminOrderController {
```
`admin`은 재웅님이 (가상으로)설립한 회사의 이 서비스 관리자인지 아니면 업체 사장님인지 정확한 정의가 필요해보입니다~
보통 주문의 상태변경은 사장님이 할 것 같습니다

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/controller/OrderController.java:
```diff
@@ -0,0 +1,75 @@
+    public ResponseEntity getOrder(@PathVariable Long orderId, @CurrentUser Account account) {
+        Order order = orderService.getOrder(orderId);
+
+        if (!Objects.equals(order.getAccount(), account)) {
```
권한관리 및 검사는 Spring security에 내장되어있습니다.
검색해보시면 좋을 것 같습니다~

jjeda replied:
제가 이해한 spring security 권한은 `ADMIN` `USER` 이런식으로 역할을 설정하여 `USER` 는 `ADMIN` 구역에 접근하지 못한다! 
이거였는데 구매자가 주문한 정보를 검색하려고 할때 내가 아닌 다른사람의 접근을 막으려고 하는 권한 검사도 가능하다는 말씀이시죠~?

f-lab-dev replied:
아~ 이쪽은 제가 잘못봤었네요. 역할검사로 착각했습니다!

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/domain/Order.java:
```diff
@@ -40,11 +44,11 @@
      * 주로 access 하는 Order를 연관관계의 주인으로 설정
      */
-    @OneToOne
+    @OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
```
[팁] FetchType에 따라서 성능이 달라질 수도 있습니다. 각각 쿼리가 어떻게 나가는지 디버깅해보시고 어떤 상황에 써야할지 고민해보시는것도 좋을 것 같네요.

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/dto/DeliveryDto.java:
```diff
@@ -0,0 +1,23 @@
+
+    private DeliveryStatus status;
+
+    public Delivery from() {
```
이 메소드는 from보다는 to가 적절해보이네요~

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/dto/OrderDto.java:
```diff
@@ -0,0 +1,41 @@
+
+    private Account account;
+
+    public Order from() {
```
여기도 `to()`가 적절해보입니다.
from(other) => 다른 객체로부터 자신의 객체를 생성
to => 자신의 객체로부터 다른 객체를 생성

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/service/OrderService.java:
```diff
@@ -0,0 +1,100 @@
+
+        // 연관관계 메서드
+        order.setAccount(account);
+        account.getOrders().add(order);
```
이렇게 하면 사용하지 않는 order의 데이터를 불러오기 위해 쿼리가 날아갈텐데 order를 insert하는 로직은 따로 두면 좋을 것 같습니다.

f-lab-dev replied:
페이징도 하지 않는 것 같아서 장애의 원인이 될 수 있을 것 같습니다

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/service/OrderService.java:
```diff
@@ -0,0 +1,100 @@
+        /* 주문이 완료되면 아이템의 전체 재고에서 주문수량만큼 빼주어야한다. */
+        for (OrderItem orderItem : orderItems) {
+            Item item = orderItem.getItem();
+            itemService.removeStock(item.getId(), orderItem.getQuantity());
```
여기선 반복문을 돌며 DB를 호출하게되는데 n+1 쿼리는 좋지 않습니다. 벌크호출을 하는게 좋아보입니다~

jjeda replied:
@f-lab-dev
이상황에서 벌크호출은 `@Modifying` + `@Query` 로 직접 쿼리를 날려서 해결하는 건가요? 
키워드좀 부탁드려도 될까요~

f-lab-dev replied:
네 이부분은 직접 쿼리를 구현해야할 것 같네요~ Repository에서 `@Query` 어노테이션을 사용해서 메소드를 구현하시면 될것같아요.
추가적으로 메소드명도 remove보단 decrementStock 같은게 좋아보이네요~

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/service/OrderService.java:
```diff
@@ -0,0 +1,100 @@
+        return orderRepository.save(order);
+    }
+
+    @Transactional(readOnly = true)
```
여기서는 굳이 트랜잭션을 걸어주지 않아도 될것같네요~

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/service/OrderService.java:
```diff
@@ -0,0 +1,100 @@
+     * 준비 -> 배송중  / 배송중 -> 완료 메서드로 분리
+     */
+    @Transactional
+    public Order startDelivery(Long orderId) {
```
Controller에 달은 리뷰와 마찬가지로 너무 메소드가 정적이라 좋지 않아보입니다~

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/service/OrderService.java:
```diff
@@ -0,0 +1,100 @@
+
+        DeliveryStatus deliveryStatus = order.getDelivery().getStatus();
+        if (Objects.equals(deliveryStatus, DeliveryStatus.DELIVERY) || Objects.equals(deliveryStatus, DeliveryStatus.COMP)) {
+            throw new IllegalStateException("배송중이거나 배송완료된 상품은 취소가 불가능합니다.");
```
👍

jjeda committed:
[#22] OrderController 분리

- Buyer / Seller 주문관련 컨트롤러 분리
- Dto 메서드 이름 수정
- 서비스 내 정적 메서드 수정

f-lab-dev commented on src/main/java/me/jjeda/mall/orders/dto/OrderDto.java:
```diff
@@ -24,14 +24,14 @@
     private Account account;
 
-    public Order from() {
+    public Order to() {
```
아 이쪽은 이름을 잘못 제시해드린것 같습니다. `toEntity` 정도가 더 적당해보입니다. 명확성이 필요해보이네요!

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/domain/Account.java:
```diff
@@ -77,4 +77,8 @@ public void update(AccountDto accountDto) {
     }
+
+    public void insertOrder(Order order) {
+        this.orders.add(order);
```
이 문장으로 인해 생기는 쿼리를 select 후에 insert하는건지 디버깅해보면 좋을 것 같습니다.
만약 위처럼 쿼리가 생성된다면 수정해주는게 좋을 것 같네요

jjeda committed:
[#22] 주문관련 테스트코드 작성

- 주문완료 테스트

jjeda commented on pom.xml:
```diff
@@ -43,18 +43,23 @@
+<!--            <groupId>mysql</groupId>-->
+<!--            <artifactId>mysql-connector-java</artifactId>-->
+<!--            <scope>runtime</scope>-->
+<!--        </dependency>-->
```
mysql 관련 오류잡는중에 주석처리하고 시도해보다가 커밋되었네요 ㅠㅠ 다음커밋에 바로 수정하겠습니다

jjeda commented on src/test/java/me/jjeda/mall/orders/controller/OrderControllerTest.java:
```diff
@@ -1,11 +1,137 @@
 
+        return "bearer " + parser.parseMap(responseBody).get("access_token").toString();
 
+    }
```
그동안 코드를 수정하면서 테스트코드들을 점검하지 않고오다보니까 많이 깨져있네요.. 하루종일 에러만잡다가 결국못잡았네요
단순히 주문생성하는 테스트를 만들던 도중에 수많은 오류들과 마주친...
기존 로직인 토큰발급하는 부분에서 아직 에러가있어 이 테스트코드는 미완성입니다(단순히 print만 찍어놨네요..)

jjeda committed:
[#22] 주문 관련 전체 로직 수정

- AccountAdapter가 AccountDto가 아닌 Account를 주입 받도록 재수정
- Redis SerializtionException 으로 인한 토큰스토어 임시변경
- 주문생성 테스트코드 수정

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/controller/AccountController.java:
```diff
@@ -37,20 +37,20 @@ public ResponseEntity createAccount(@RequestBody @Valid AccountDto requestAccoun
     @GetMapping
-    public ResponseEntity getAccount(@CurrentUser Account account) {
-        return ResponseEntity.ok(account);
+    public ResponseEntity getAccount(@CurrentUser AccountDto accountDto) {
```
매번 호출마다 DB에서 정보를 받아올 필요는 없어보입니다. userId정도만 받아와서 User 객체가 필요하면 그때그때 쿼리를 날리는게 좋을 것 같네요.

f-lab-dev commented on src/main/java/me/jjeda/mall/common/configs/ModelMapperConfig.java:
```diff
@@ -0,0 +1,14 @@
+public class ModelMapperConfig {
+
+    @Bean
+    public ModelMapper modelMapper() {
```
이 ModelMapper 객체는 thread safe인지 알아보는게 좋을 것 같네요.

jjeda committed:
[#30] 주문상품에 상품담기

- 상품관련 테스트

jjeda committed:
[#30] merge branch

- #33 branch -> #30

jjeda committed:
[#30] for문 -> stream 수정

jjeda committed:
Merge pull request #34 from f-lab-edu/feature/30

[#30] 주문상품에 상품담기

jjeda committed:
[#35] OrderService unit test

- 정상적인 주문 생성 단위테스트
- 주문관련 도메인 Getter 에 @JsonIgnore 처리

jjeda committed:
[#35] OrderService Unit Test

- 기존 단위테스트 수정
- 서비스 레이어 단위테스트 보완

f-lab-dev commented:
이 PR에서 받은 코드리뷰를 해결하는 것은 동일한 브랜치에서 해야합니다~
그래야 리뷰한 내용이 수정되었는지 보기 쉬워집니다.
리뷰내용 수정되었으면 머지해도 될 것 같습니다~

jjeda commented:
> 이 PR에서 받은 코드리뷰를 해결하는 것은 동일한 브랜치에서 해야합니다~
> 그래야 리뷰한 내용이 수정되었는지 보기 쉬워집니다.
> 리뷰내용 수정되었으면 머지해도 될 것 같습니다~

jjeda commented:
아하 그렇다면 상위 브랜치에서 수정후 하위 브랜치에서 수정내용을 받아오기위해
`git merge` 로 받아오는식으로 작업하면 될까요?

jjeda committed:
[#35] AccountService 유닛테스트 수정

- 정상적으로 주문생성하는 테스트 수정
- AccountController -> AccountService 에서 Entity 가아닌 Dto를 받게 수정

jjeda committed:
test automation 테스트

jjeda committed:
[#13] Account 은닉화

- AccountService 에서 Entity가 아닌 Dto를 리턴하도록 수정
- AccountController에서 Entity 를 받아오는 것이 아닌 Dto를받아오게 수정
- Account <-> AccountDto Circular Dependency 문제해결을 위한 AccountAndDtoAdapter 적용
- Session에 저장할때 Account가 아닌 AccountDto를 저장하도록 수정
- 현재사용자를 받아올때 account를 받아오는것이아닌 AccountDto를 받아오도록 수정
- 테스트코드 수정

jjeda committed:
[#13] Account <-> Dto 어댑터 수정

- DtoToEntity 메서드 -> dtoToEntity
- 메서드 중복 로직 제거

jjeda committed:
Merge pull request #37 from f-lab-edu/feature/13

[#13] Account 은닉화

jjeda committed:
Merge pull request #36 from f-lab-edu/feature/35

[#35] OrderService unit test
